engine/src/main/java/org/hibernate/search/engine/mapper/schema/management/SchemaElementContext.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.engine.mapper.schema.management;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Optional;
+
+import org.hibernate.search.util.common.annotation.Incubating;
+
+@Incubating
+public interface SchemaElementContext {
+	enum OutputFormat {
+		/**
+		 * Generate a JSON representation of this context.
+		 */
+		SIMPLE;
+	}
+
+	/**
+	 * @return The name of the backend to which the index this context represents belongs to.
+	 * If it is a part of the default backend - {@link Optional#empty() an empty optional} will be returned.
+	 */
+	Optional<String> backendName();
+
+	/**
+	 * @return The name of the index this context represents.
+	 */
+	String indexName();
+
+	/**
+	 * Writes the content of this context to provided output stream based on a requested format.
+	 *
+	 * @param format The format in which the output should be represented.
+	 * @param stream The stream to which the content should be written.
+	 * @throws IOException If any {@link IOException I/O problems} occur during the process.
+	 * @see OutputFormat
+	 */
+	void write(OutputFormat format, OutputStream stream) throws IOException;
+
+	/**
+	 * Unwraps this context to a more specific (backend) version.
+	 */
+	<T extends SchemaElementContext> T unwrap(Class<T> klass);
```

Comment:
Would you mind using `ElasticsearchExtension` instead? I've been kind of encouraged to expose `unwrap` methods in the past, but I definitely don't like it...

The nice thing about `ElasticsearchExtension` is that:

1. It's discoverable; you don't need to know the name of the Elasticsearch-specific class, you just pass `ElasticsearchExtension.get()` and automatically get the correct type as a result.
2. It works with parameterized types, though here that's not important.

See for example `org.hibernate.search.backend.elasticsearch.search.query.impl.ElasticsearchSearchQueryImpl#extension`.